base_layer/core/src/validation/block_validators.rs

```diff
@@ -63,23 +68,41 @@ impl Validation<Block> for StatelessBlockValidator {
     /// 1. Is there precisely one Coinbase output and is it correctly defined with the correct amount?
     /// 1. Is the accounting correct?
     fn validate(&self, block: &Block) -> Result<(), ValidationError> {
-        let block_id = format!("block #{} ({})", block.header.height, block.hash().to_hex());
-        check_block_weight(block, &self.rules.consensus_constants(block.header.height))?;
+        if block.header.height == 0 {
+            warn!(target: LOG_TARGET, "Attempt to validate genesis block");
+            return Err(ValidationError::ValidatingGenesis);
+        }
+
+        let block_id = if cfg!(debug_assertions) {
+            format!("block #{} ({})", block.header.height, block.hash().to_hex())
+        } else {
+            format!("block #{}", block.header.height)
+        };
+        trace!(target: LOG_TARGET, "Validating {}", block_id);
+
+        helpers::check_block_weight(&block, &self.rules.consensus_constants(block.header.height))?;
         trace!(target: LOG_TARGET, "SV - Block weight is ok for {} ", &block_id);
-        check_duplicate_transactions_inputs(block)?;
+
+        trace!(
+            target: LOG_TARGET,
+            "Checking duplicate inputs and outputs on {}",
+            block_id
+        );
+        check_duplicate_transactions_inputs(&block.body)?;
         trace!(
             target: LOG_TARGET,
             "SV - No duplicate inputs or outputs for {} ",
             &block_id
         );
+
         // Check that the inputs are are allowed to be spent
         block.check_stxo_rules()?;
         trace!(target: LOG_TARGET, "SV - Output constraints are ok for {} ", &block_id);
         check_cut_through(block)?;
         trace!(target: LOG_TARGET, "SV - Cut-through is ok for {} ", &block_id);
-        check_coinbase_output(block, self.rules.clone(), &self.factories)?;
+        check_coinbase_output(block, &self.rules, &self.factories)?;
         trace!(target: LOG_TARGET, "SV - Coinbase output is ok for {} ", &block_id);
-        check_accounting_balance(block, self.rules.clone(), &self.factories)?;
+        check_accounting_balance(block, &self.rules, &self.factories)?;
```

Comment:
You can safely and cheaply clone consensusmanager, its an arc.

Reply:
Passing a reference does not require an atomic update of the Arc's counter.